lib/blob.js

```diff
@@ -0,0 +1,377 @@
+'use strict';
+
+var assert            = require('assert');
+var crypto            = require('crypto');
+var debug             = require('debug')('azure:blob');
+var Promise           = require('promise');
+var utils             = require('./utils');
+var querystring       = require('querystring');
+var xml               = require('./xml-parser');
+
+/*
+ * Azure storage service version
+ * @const
+ */
+var SERVICE_VERSION = '2015-12-11';
+
+/* Transient error codes (we'll retry request when encountering these codes */
+var TRANSIENT_ERROR_CODES = [
+  // Azure error codes we should retry on according to azure docs
+  'InternalError',
+  'ServerBusy'
+].concat(utils.TRANSIENT_HTTP_ERROR_CODES);
+
+/*
+ * List of query-string parameter supported in lexicographical order, used for
+ * construction of the canonicalized resource.
+ */
+var QUERY_PARAMS_SUPPORTED = [
+  'comp',
+  'timeout',
+  'restype'
+].sort();
+
+/*
+ * Authorize the request with shared key
+ * Intended to define `Blob.prototype.authorize`.
+ */
+var authorizeWithSharedKey = function (method, path, query, headers) {
```

Comment:
It seems like we have almost the exact same function in `lib/{blob,queue,table}.js`.  I suspect that the simplest version was done first for Table storage, then expanded to cover Queue storage as we added that support.  I think that instead of implementing it in three slightly different ways, we might want to have a single underlying implementation.  I think it's reasonable to have a shared underlying implementation.  This could take the `method, path, query, headers` parameters as well as a list of things like the headers to include in the signing and list of supported query string params.  We could then have specializations of the `.authorizeWithSharedKey()` function which sort of look like:

```javascript
var authorizeWithSharedKey = function (m, p, q, h) {
  return utils.buildAuthDoohicky(headersToLookAt, supportedQueryParams, m, p, q, h);
}
```

What do you think?  Maybe the first pass could be making the general methods and implementing them for Blob, leaving Queue and Table for when we're sure the general method works.